lib/app/features/nostr/providers/nostr_notifier.c.dart

```diff
@@ -72,9 +77,43 @@ class NostrNotifier extends _$NostrNotifier {
     RequestMessage requestMessage, {
     ActionSource actionSource = const ActionSourceCurrentUser(),
   }) async* {
-    //TODO: handle NOTICE
     final relay = await _getRelay(actionSource);
-    yield* nd.requestEvents(requestMessage, relay);
+
+    final dislikedRelaysUrls = <String>[];
+    StreamSubscription<RelayMessage>? subscription;
+    final streamController = StreamController<EventMessage>(onCancel: () => subscription?.cancel());
+
+    Future<void> handleRetry() => withRetry(
+          () async {
+            final retryRelay = await _getRelay(actionSource, dislikedUrls: dislikedRelaysUrls);
+            final retryEventsStream = nd.requestEvents(requestMessage, retryRelay);
+            await for (final retryEvent in retryEventsStream) {
+              if (streamController.isClosed) break;
+              if (retryEvent.indicatesError) {
+                dislikedRelaysUrls.add(retryRelay.url);
+                throw RelayRequestFailedException();
+              } else if (retryEvent is EventMessage) {
+                streamController.add(retryEvent);
+              }
+            }
+          },
+        );
+
+    subscription = nd.requestEvents(requestMessage, relay).listen(
+      (event) async {
+        if (event.indicatesError) {
+          await subscription?.cancel();
+          dislikedRelaysUrls.add(relay.url);
+          await handleRetry();
```

Comment:
it's more complicated here, a retry is a response to a specific stream event. We cannot just wrap it with `handleRetry` because it will not wait for all the events from a stream to be fired.

Reply:
that's brilliant actually, I agree.